src/main/scala/cognite/spark/v1/SequencesRelation.scala

```diff
@@ -91,43 +141,84 @@ class SequencesRelation(config: RelationConfig)(val sqlContext: SQLContext)
       .flatTap(_ => incMetrics(itemsDeleted, ids.length))
   }
 
-  override def upsert(rows: Seq[Row]): IO[Unit] =
-    // Sequences fail with 400 error code when id already exists:
-    // /api/v1/projects/jetfiretest2/sequences failed with status 400: The following external id(s) are already in the database:
-    throw new CdfSparkException("Upsert not supported for sequences.")
+  override def upsert(rows: Seq[Row]): IO[Unit] = {
+    val sequences =
+      rows
+        .map(fromRow[SequenceUpsertSchema](_))
+
+    implicit val toCreate =
+      Transformer
+        .define[SequenceUpsertSchema, SequenceCreate]
+        .withFieldComputed(
+          _.columns,
+          x => x.getSequenceColumnCreate
+        )
+        .buildTransformer
+
+    val existingSeqs: Map[Option[Long], Seq[String]] = getExistingSequenceColumnsByIds(sequences)
+    val existingSeqsWithExtId: Map[Option[String], Seq[String]] =
+      getExistingSequenceColumnsByExternalIds(sequences)
 
+    implicit val toUpdate = Transformer
+      .define[SequenceUpsertSchema, SequenceUpdate]
+      .withFieldComputed(
+        _.columns,
+        x =>
+          x.getSequenceColumnsUpdate(
+            existingSeqs
+              .getOrElse(x.id, existingSeqsWithExtId.getOrElse(x.externalId.toOption, Seq()))
+              .toSet)
+      )
+      .buildTransformer
+
+    genericUpsert[Sequence, SequenceUpsertSchema, SequenceCreate, SequenceUpdate, SequencesResource[IO]](
+      sequences,
+      isUpdateEmpty,
+      client.sequences
+    )
+  }
+
+  // scalastyle:off method.length
   override def getFromRowsAndCreate(rows: Seq[Row], doUpsert: Boolean = true): IO[Unit] = {
     val sequences =
       rows
-        .map(fromRow[SequenceUpdateSchema](_))
+        .map(fromRow[SequenceUpsertSchema](_))
 
     implicit val toCreate =
       Transformer
-        .define[SequenceUpdateSchema, SequenceCreate]
+        .define[SequenceUpsertSchema, SequenceCreate]
```

Comment:
I didn't like defining these implicit transformers inside functions multiple times but they are dynamically defined according to the list of columns of sequences. And moving outside does not work. Let me know if there is a nicer way.

Reply:
It just need to find the implicit val, so you can define those as private def and declare the implicit val inside each bloc to retrieve the value from those private def 😁 , sorry I wanted to create a PR to your branch but push in the wrong same branch 🤦